cam6_2_021: Enhancements to wet removal of aerosols

fvitt opened the pull request:
These new features are turned off by default.  Baselines should be bit-for-bit unchanged.

closes issue #91
closes issue #92 -- bug fix for noleap calendar with ESMF lib
closes PR #99

nusbaume commented on src/chemistry/modal_aero/modal_aero_convproc.F90:
```diff
@@ -108,6 +120,58 @@ subroutine ma_convproc_register
+  end if
+
+  ! Broadcast namelist variables
+  call mpi_bcast( convproc_do_gas,  1, mpi_logical, masterprocid, mpicom, ierr)
```
Should these calls have error checks, for example:

`if (ierr /= 0) call endrun(sub//": FATAL: mpi_bcast: convproc_do_gas")`

I realize the odds of this happening are low, but it couldn't hurt.

gold2718 replied:
I've gone back and forth on this over the decades. I used to think that catching stuff like this* was a good idea. On the other hand, in 2020, an error in a place like this will crash and burn with little likelihood that control will ever return to process this check.
Now, for fancy MPI code (e.g., asynchronous sends and receives), it is often worth trying to catch these errors but I now feel that error checks in places like this is a waste of code. To be clear, this is because of the evolution of network fabric (e.g., Infiniband) and MPI libraries, in general, it is good SE practice.

*Note that, by default, MPI routines such as MPI_bcast will not return on error. You have to set an error handler for that.

fvitt replied:
If checking for error is a waste of code then I will leave as it is, if there are no objections.

nusbaume replied:
That's fine with me, given @gold2718's explanation.  I'll go ahead and resolve this conversation.

gold2718 commented on src/chemistry/aerosol/wetdep.F90:
```diff
@@ -447,6 +475,13 @@ subroutine wetdepa_v2(                                  &
+         ! If resuspending aerosol only when all the rain has totally
+         ! evaporated then zero out any aerosol tendency for partial
+         ! evaporation.
+         if (convproc_do_evaprain_atonce .and. (evaps(i,k)*pdog(i).ne.precabs(i))) then
```
Still using `.ne.` over `/=`?

fvitt replied:
see below

gold2718 commented on src/chemistry/modal_aero/aero_model.F90:
```diff
@@ -1603,7 +1613,36 @@ subroutine aero_model_wetdep( state, dt, dlf, cam_out, ptend, pbuf)
+                         mm = lmassptrcw_amode(lspec,m)
+                      endif
+                   endif
+                   if (lphase .eq. 2) then
```
Please be consistent (e.g., you* use '==' on line 1624).

*I have no idea where this code came from so I'm blaming you :)

fvitt replied:
Many hands have modified this code so you see different styles.. I tried to make the styles of conditional operators consistent.

gold2718 commented on src/chemistry/modal_aero/modal_aero_convproc.F90:
```diff
@@ -108,6 +120,58 @@ subroutine ma_convproc_register
+  ! Read namelist
+  if (masterproc) then
+     unitn = getunit()
+     open( unitn, file=trim(nlfile), status='old' )
```
`getunit` and `freeunit` are not needed and should be deprecated. Please get rid of them and use:
```
open( newunit=unitn, file=trim(nlfile), status='old' )
```

fvitt replied:
OK good to know.  I made the change as you suggest.

cacraigucar commented on src/chemistry/modal_aero/modal_aero_convproc.F90:
```diff
@@ -1143,6 +1213,15 @@ subroutine ma_convproc_tend(                                           &
+! in the first version of this routine, the precp evap resusp tendencies for k > kbot were ignored, 
+!    but that is now fixed
+! this was a minor bug with quite minor affects on the aerosol, 
+!    because convective precip evap is (or used to be) much less than stratiform precip evap )
```
This is a lengthy comment of what "used to be", including the developers name and a date over 4 years ago?

Should some (all?) of this comment be moved to the ChangeLog?

fvitt replied:
I moved some of this to ChangeLog.  Please review.

fvitt commented:
I believe I have addressed the reviewer's issues.  I have included a small change to src/chemistry/mozart/mo_extfrc.F90 to improve error checking on input files.  Please review.
